// sub-agents-mcp/dist/utils/Logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = void 0;
const node_fs_1 = require("node:fs");
const node_path_1 = require("node:path");
/**
 * Logger utility class for structured logging.
 *
 * Provides consistent logging format with timestamps, levels, and context data.
 * Supports filtering by log level and structured output for debugging.
 */
class Logger {
    constructor(level = 'info') {
        this.fileWriteErrorShown = false;
        this.currentLevel = level;
        // Check for optional file logging via environment variable
        const logFile = process.env['LOG_FILE'];
        if (logFile?.trim()) {
            this.logFilePath = logFile.trim();
        }
    }
    /**
     * Logs a debug message.
     *
     * @param message - The message to log
     * @param context - Optional context data
     */
    debug(message, context) {
        this.log('debug', message, context);
    }
    /**
     * Logs an info message.
     *
     * @param message - The message to log
     * @param context - Optional context data
     */
    info(message, context) {
        this.log('info', message, context);
    }
    /**
     * Logs a warning message.
     *
     * @param message - The message to log
     * @param context - Optional context data
     */
    warn(message, context) {
        this.log('warn', message, context);
    }
    /**
     * Logs an error message.
     *
     * @param message - The message to log
     * @param error - Optional error object
     * @param context - Optional context data
     */
    error(message, error, context) {
        const errorContext = error ? { ...context, error: error.message, stack: error.stack } : context;
        this.log('error', message, errorContext);
    }
    /**
     * Internal log method that handles the actual logging.
     *
     * @param level - The log level
     * @param message - The message to log
     * @param context - Optional context data
     */
    log(level, message, context) {
        if (!this.shouldLog(level)) {
            return;
        }
        const logEntry = {
            timestamp: new Date().toISOString(),
            level,
            message,
            ...(context && { context }),
        };
        this.output(logEntry);
    }
    /**
     * Determines if a message should be logged based on current level.
     *
     * @param level - The log level to check
     * @returns True if the message should be logged
     */
    shouldLog(level) {
        return Logger.levelPriority[level] >= Logger.levelPriority[this.currentLevel];
    }
    /**
     * Outputs the log entry to the console and optionally to a file.
     *
     * @param entry - The log entry to output
     */
    output(entry) {
        const { timestamp, level, message, context } = entry;
        // Format output for console readability
        const formattedMessage = `[${timestamp}] ${level.toUpperCase()}: ${message}`;
        if (context && Object.keys(context).length > 0) {
            console.error(formattedMessage, context);
        }
        else {
            console.error(formattedMessage);
        }
        // Optional file output in JSON format
        if (this.logFilePath) {
            this.writeToFile(entry);
        }
    }
    /**
     * Writes log entry to file in JSON format.
     * Errors are handled internally and do not affect main logging flow.
     *
     * @param entry - The log entry to write to file
     */
    writeToFile(entry) {
        // Fire-and-forget async operation to avoid blocking main thread
        this.performFileWrite(entry).catch((error) => {
            // Only show file write error once to avoid spam
            if (!this.fileWriteErrorShown) {
                this.fileWriteErrorShown = true;
                console.error('Logger: Failed to write to log file, falling back to console only', error);
            }
        });
    }
    /**
     * Performs the actual file write operation.
     *
     * @param entry - The log entry to write
     */
    async performFileWrite(entry) {
        if (!this.logFilePath)
            return;
        // Ensure directory exists
        const dir = (0, node_path_1.dirname)(this.logFilePath);
        await node_fs_1.promises.mkdir(dir, { recursive: true });
        // Convert entry to JSON string
        const jsonLine = `${JSON.stringify(entry)}\n`;
        // Append to file
        await node_fs_1.promises.appendFile(this.logFilePath, jsonLine, 'utf8');
    }
}
exports.Logger = Logger;
/**
 * Log level priority for filtering.
 */
Logger.levelPriority = {
    debug: 0,
    info: 1,
    warn: 2,
    error: 3,
};
