// sub-agents-mcp/dist/execution/McpRequestTimeout.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.McpRequestTimeout = exports.DEFAULT_MCP_TIMEOUT_CONFIG = void 0;
const Logger_1 = require("../utils/Logger");
/**
 * Default timeout configuration for MCP requests
 * Based on MCP best practices and LLM operation characteristics
 */
exports.DEFAULT_MCP_TIMEOUT_CONFIG = {
    defaultTimeoutMs: 360000, // 6 minutes - considering complex agent operations
    maxTimeoutMs: 660000, // 11 minutes - absolute maximum for AI->MCP
    progressResetEnabled: true,
    warningThresholdMs: 60000, // 1 minute - send progress notification
    enableDebugLogging: false,
};
/**
 * McpRequestTimeout manages timeout behavior for MCP server requests
 * Implements MCP specification recommendations for handling long-running operations
 *
 * This class handles AI -> MCP level timeouts, separate from MCP -> AI (AgentExecutor) timeouts
 * Following MCP best practices:
 * - Progressive timeout with reset on progress
 * - Maximum absolute timeout enforcement
 * - Progress notifications for better UX
 * - Graceful cancellation support
 */
class McpRequestTimeout {
    /**
     * Creates a new McpRequestTimeout instance
     *
     * @param config - Optional timeout configuration
     * @param logger - Optional logger instance
     */
    constructor(config, logger) {
        this.contexts = new Map();
        this.config = { ...exports.DEFAULT_MCP_TIMEOUT_CONFIG, ...config };
        this.logger = logger || new Logger_1.Logger(process.env['LOG_LEVEL'] || 'info');
    }
    /**
     * Start timeout tracking for a request
     *
     * @param requestId - Unique request identifier
     * @param onTimeout - Callback when timeout occurs
     * @param onProgress - Optional callback for progress notifications
     * @returns TimeoutContext for the request
     */
    startTimeout(requestId, onTimeout, onProgress) {
        // Clear any existing timeout for this request
        this.clearTimeout(requestId);
        const context = {
            requestId,
            startTime: Date.now(),
            currentTimeout: null,
            warningTimeout: null,
            maxTimeout: null,
            progressCount: 0,
            lastProgressTime: Date.now(),
            isCompleted: false,
            isCancelled: false,
            onTimeout,
            ...(onProgress && { onProgress }),
        };
        // Set warning timeout (for progress notification)
        if (onProgress && this.config.warningThresholdMs > 0) {
            context.warningTimeout = setTimeout(() => {
                if (!context.isCompleted && !context.isCancelled) {
                    this.sendProgressNotification(context, onProgress, 'Processing is taking longer than expected');
                }
            }, this.config.warningThresholdMs);
        }
        // Set default timeout
        context.currentTimeout = setTimeout(() => {
            if (!context.isCompleted && !context.isCancelled) {
                this.handleTimeout(context, onTimeout);
            }
        }, this.config.defaultTimeoutMs);
        // Set maximum absolute timeout (cannot be reset)
        context.maxTimeout = setTimeout(() => {
            if (!context.isCompleted && !context.isCancelled) {
                this.logger.warn('Maximum timeout reached, forcing termination', {
                    requestId,
                    elapsedMs: Date.now() - context.startTime,
                    maxTimeoutMs: this.config.maxTimeoutMs,
                });
                this.handleTimeout(context, onTimeout);
            }
        }, this.config.maxTimeoutMs);
        this.contexts.set(requestId, context);
        if (this.config.enableDebugLogging) {
            this.logger.debug('Timeout tracking started', {
                requestId,
                defaultTimeoutMs: this.config.defaultTimeoutMs,
                maxTimeoutMs: this.config.maxTimeoutMs,
            });
        }
        return context;
    }
    /**
     * Report progress and optionally reset timeout
     * Following MCP specification: MAY reset timeout on progress, but enforce maximum
     *
     * @param requestId - Request identifier
     * @param message - Progress message
     * @param percentage - Optional completion percentage
     */
    reportProgress(requestId, message, percentage) {
        const context = this.contexts.get(requestId);
        if (!context || context.isCompleted || context.isCancelled) {
            return;
        }
        context.progressCount++;
        context.lastProgressTime = Date.now();
        if (this.config.enableDebugLogging) {
            this.logger.debug('Progress reported', {
                requestId,
                message,
                percentage,
                progressCount: context.progressCount,
                elapsedMs: Date.now() - context.startTime,
            });
        }
        // Reset timeout if enabled (but maximum timeout remains)
        if (this.config.progressResetEnabled && context.currentTimeout) {
            clearTimeout(context.currentTimeout);
            // Calculate remaining time before max timeout
            const elapsedMs = Date.now() - context.startTime;
            const remainingMs = this.config.maxTimeoutMs - elapsedMs;
            const newTimeoutMs = Math.min(this.config.defaultTimeoutMs, remainingMs);
            if (newTimeoutMs > 0 && context.onTimeout) {
                context.currentTimeout = setTimeout(() => {
                    if (!context.isCompleted && !context.isCancelled) {
                        this.handleTimeout(context, context.onTimeout);
                    }
                }, newTimeoutMs);
                this.logger.info('Timeout reset due to progress', {
                    requestId,
                    newTimeoutMs,
                    remainingBeforeMaxMs: remainingMs,
                });
            }
        }
    }
    /**
     * Mark request as completed and clear timeouts
     *
     * @param requestId - Request identifier
     */
    complete(requestId) {
        const context = this.contexts.get(requestId);
        if (!context) {
            return;
        }
        context.isCompleted = true;
        this.clearTimeouts(context);
        const totalTime = Date.now() - context.startTime;
        this.logger.info('Request completed within timeout', {
            requestId,
            totalTimeMs: totalTime,
            progressCount: context.progressCount,
        });
        this.contexts.delete(requestId);
    }
    /**
     * Cancel request and clear timeouts
     * Following MCP specification for cancellation
     *
     * @param requestId - Request identifier
     * @param reason - Cancellation reason
     */
    cancel(requestId, reason) {
        const context = this.contexts.get(requestId);
        if (!context) {
            return;
        }
        context.isCancelled = true;
        this.clearTimeouts(context);
        this.logger.info('Request cancelled', {
            requestId,
            reason,
            elapsedMs: Date.now() - context.startTime,
        });
        this.contexts.delete(requestId);
    }
    /**
     * Clear timeout for a request
     *
     * @param requestId - Request identifier
     */
    clearTimeout(requestId) {
        const context = this.contexts.get(requestId);
        if (context) {
            this.clearTimeouts(context);
            this.contexts.delete(requestId);
        }
    }
    /**
     * Check if request has timed out
     *
     * @param requestId - Request identifier
     * @returns true if timed out
     */
    hasTimedOut(requestId) {
        const context = this.contexts.get(requestId);
        if (!context) {
            return false;
        }
        const elapsedMs = Date.now() - context.startTime;
        return elapsedMs >= this.config.maxTimeoutMs;
    }
    /**
     * Get timeout statistics for monitoring
     *
     * @returns Current timeout statistics
     */
    getStats() {
        const activeContexts = Array.from(this.contexts.values()).filter((c) => !c.isCompleted && !c.isCancelled);
        const now = Date.now();
        const longestRunningMs = activeContexts.reduce((max, c) => Math.max(max, now - c.startTime), 0);
        const averageProgressCount = activeContexts.length > 0
            ? activeContexts.reduce((sum, c) => sum + c.progressCount, 0) / activeContexts.length
            : 0;
        return {
            activeRequests: activeContexts.length,
            averageProgressCount,
            longestRunningMs,
        };
    }
    /**
     * Handle timeout event
     *
     * @private
     */
    handleTimeout(context, onTimeout) {
        const elapsedMs = Date.now() - context.startTime;
        this.logger.warn('Request timeout occurred', {
            requestId: context.requestId,
            elapsedMs,
            progressCount: context.progressCount,
            lastProgressMs: Date.now() - context.lastProgressTime,
        });
        onTimeout(context);
        this.clearTimeouts(context);
    }
    /**
     * Send progress notification
     *
     * @private
     */
    sendProgressNotification(context, onProgress, message) {
        const notification = {
            requestId: context.requestId,
            message,
            timestamp: Date.now(),
        };
        onProgress(notification);
    }
    /**
     * Clear all timeouts for a context
     *
     * @private
     */
    clearTimeouts(context) {
        if (context.currentTimeout) {
            clearTimeout(context.currentTimeout);
            context.currentTimeout = null;
        }
        if (context.warningTimeout) {
            clearTimeout(context.warningTimeout);
            context.warningTimeout = null;
        }
        if (context.maxTimeout) {
            clearTimeout(context.maxTimeout);
            context.maxTimeout = null;
        }
    }
}
exports.McpRequestTimeout = McpRequestTimeout;
