// sub-agents-mcp/dist/execution/PromptController.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PromptController = exports.DEFAULT_RECURSION_WARNING_CONFIG = void 0;
/**
 * Default configuration for recursion prevention warnings.
 * Uses clear English instructions to prevent sub-agents-mcp MCP tool usage.
 */
exports.DEFAULT_RECURSION_WARNING_CONFIG = {
    warningTemplate: `
IMPORTANT RECURSION PREVENTION INSTRUCTIONS:
- Do NOT use sub-agents-mcp MCP tools in your response
- Complete this task within the current agent context only
- Do not attempt to call other agents or sub-agents
- Focus on direct implementation and solution delivery
`,
    useFormatting: true,
};
/**
 * Controller class for enhancing prompts with recursion prevention instructions.
 * Implements the security strategy defined in ADR-0001 for preventing infinite loops
 * in agent-to-agent communication through prompt control.
 */
class PromptController {
    /**
     * Creates a new PromptController instance.
     *
     * @param config - Optional configuration for recursion warning behavior
     */
    constructor(config = exports.DEFAULT_RECURSION_WARNING_CONFIG) {
        this.config = config;
    }
    /**
     * Adds recursion prevention warning to a prompt string.
     *
     * This method implements the prompt control strategy defined in ADR-0001
     * to prevent infinite loops in agent-to-agent communication. It preserves
     * the original prompt content while adding clear security instructions
     * that explicitly prohibit the use of sub-agents-mcp MCP tools.
     *
     * @param originalPrompt - The original user prompt to enhance
     * @returns Enhanced prompt with recursion prevention instructions appended
     * @throws {Error} When originalPrompt is empty or contains only whitespace
     *
     * @example
     * ```typescript
     * const controller = new PromptController()
     * const enhanced = controller.addRecursionWarning("Help me debug this code")
     * console.log(enhanced)
     * // Output: "Help me debug this code\n============...===========\nIMPORTANT RECURSION PREVENTION INSTRUCTIONS:\n- Do NOT use sub-agents-mcp MCP tools..."
     * ```
     */
    addRecursionWarning(originalPrompt) {
        // Input validation: ensure prompt is not empty or whitespace-only
        if (typeof originalPrompt !== 'string') {
            throw new Error('Original prompt must be a string');
        }
        if (originalPrompt.trim().length === 0) {
            throw new Error('Original prompt cannot be empty');
        }
        // Format warning based on configuration
        const warning = this.formatWarningMessage();
        // Concatenate original prompt with warning
        return `${originalPrompt}${warning}`;
    }
    /**
     * Formats the warning message according to the current configuration.
     *
     * @private
     * @returns Formatted warning message with or without visual separators
     */
    formatWarningMessage() {
        if (this.config.useFormatting) {
            const separator = '='.repeat(60);
            return `\n${separator}\n${this.config.warningTemplate}\n${separator}\n`;
        }
        return this.config.warningTemplate;
    }
    /**
     * Enhances ExecutionParams by adding recursion prevention to the prompt field.
     *
     * This method is the primary interface for integrating prompt enhancement
     * into the agent execution pipeline. It creates a new ExecutionParams object
     * with an enhanced prompt while preserving all other execution parameters.
     * The enhanced prompt includes ADR-0001 compliant security instructions.
     *
     * @param params - Original execution parameters to enhance
     * @returns New ExecutionParams object with enhanced prompt field
     * @throws {Error} When params is invalid or prompt enhancement fails
     *
     * @example
     * ```typescript
     * const controller = new PromptController()
     * const original = {
     *   agent: "code-helper",
     *   prompt: "Fix this bug",
     *   cwd: "/project",
     *   extra_args: ["--verbose"]
     * }
     * const enhanced = controller.enhanceExecutionParams(original)
     * console.log(enhanced.prompt) // Original prompt + recursion prevention
     * console.log(enhanced.agent)  // "code-helper" (unchanged)
     * ```
     */
    enhanceExecutionParams(params) {
        // Input validation: ensure params is valid ExecutionParams object
        if (!params || typeof params !== 'object') {
            throw new Error('ExecutionParams must be a valid object');
        }
        if (!params.agent || typeof params.agent !== 'string') {
            throw new Error('ExecutionParams must have a valid agent field');
        }
        if (!params.prompt || typeof params.prompt !== 'string') {
            throw new Error('ExecutionParams must have a valid prompt field');
        }
        try {
            // Enhance the prompt with recursion prevention instructions
            const enhancedPrompt = this.addRecursionWarning(params.prompt);
            // Return new object with enhanced prompt, preserving all other fields
            return {
                ...params,
                prompt: enhancedPrompt,
            };
        }
        catch (error) {
            throw new Error(`Failed to enhance execution parameters: ${error instanceof Error ? error.message : 'Unknown error'}`);
        }
    }
}
exports.PromptController = PromptController;
